torch/distributed/distributed_c10d.py

```diff
@@ -1284,11 +1284,16 @@ def _new_process_group_helper(
                 assert isinstance(
                     pg_options, ProcessGroupNCCL.Options
                 ), "Expected pg_options argument to be of type ProcessGroupNCCL.Options"
+                if pg_options._timeout:
+                    raise ValueError(
+                        "pg_options._timeout was specified, "
+                        "but timeout kwarg has a default value that will always override it. "
+                    )
```

Comment:
Why does this option only matter under NCCL? how does this interact w/ base_pg_options._timeout?

Reply:
well, gloo/mpi paths actually use this other object called 'base_pg_options' which has its _OWN_ timeout field.  That's a whole 'nother problem, which i kinda want to try fixing too.

it looks like if someone passes in a 'pg_options' object, it only gets used for nccl.  and if you use the gloo/mpi backends, the 'base_pg_options' thing gets used.  pretty sketchy.

Reply:
note: if we do fix this, we might be blocked on the fact that ProcessGroupNCCL::Options is bound to python differently than ProcessGroup::Options.  that is also something to fix.

Reply:
One reason for the weirdness is that it always used to be `init_pg` -> creates 1 backend. Now `init_pg` -> can create multiple backends. However, `pg_options` in `init_pg` is only for NCCL and no other backends use it.

Hence all the other backends are using base_pg_options._timeout and this is not exposed to the user. We are just using the options as a property bag for the timeout value and backend name.